File: ocaml/include/graph/max_flow.ml
type 't edge = { s : int; t : int; cap : 't ref; rev : 't ref }
type 't graph = 't edge list array

let add_edge g s t cap zero =
  let cap = ref cap in
  let rev = ref zero in
  g.(s) <- { s = s; t = t; cap = cap; rev = rev } :: g.(s);
  g.(t) <- { s = t; t = s; cap = rev; rev = cap } :: g.(t)

let max_flow g s t zero inf plus minus =
  let v = Array.length g in
  let rec loop res =
    let d = Array.make v (-1) in
    let que = Queue.create () in
    d.(s) <- 0;
    Queue.push s que;
    while not (Queue.is_empty que) do
      let v = Queue.pop que in
      let update e =
        if !(e.cap) > zero && d.(e.t) < 0
        then (d.(e.t) <- d.(v) + 1; Queue.push e.t que) in
      List.iter update g.(v)
    done;
    if d.(t) < 0
    then res
    else let iter = Array.copy g in
         let rec augument v f =
           if v = t
           then f
           else let rec iter_loop () =
                  ( match iter.(v) with
                    | [] -> zero
                    | e :: new_iter ->
                       iter.(v) <- new_iter;
                       if !(e.cap) > zero && d.(v) < d.(e.t)
                       then let ff = augument e.t (min f !(e.cap)) in
                            if ff > zero
                            then ( e.cap := minus !(e.cap) ff;
                                   e.rev := plus !(e.rev) ff;
                                   ff )
                            else iter_loop ()
                       else iter_loop () ) in
                iter_loop () in
         let rec loop_augument res =
           let f = augument s inf in
           if f > zero then loop_augument (plus res f) else res in
         loop (loop_augument res) in
  loop zero

File: cpp/include/graph/max_flow.cpp
#pragma once

#include "../template/const_value.hpp"
#include "definition.hpp"

template <typename edge_t, typename cap_type = typename edge_t::capacity_type>
cap_type augment(graph_t<edge_t> &g, std::vector<int> &d,
                 std::vector<int> &iter, int v, int t, const cap_type &f) {
  if (v == t) {
    return f;
  }
  for (int &i = iter[v]; i < (int)g[v].size(); i++) {
    auto &e = g[v][i];
    if (e.cap > zero<cap_type>() && d[v] < d[e.to]) {
      cap_type ff = augment(g, d, iter, e.to, t, std::min(f, e.cap));
      if (ff > zero<cap_type>()) {
        e.cap -= ff;
        g[e.to][e.rev].cap += ff;
        return ff;
      }
    }
  }
  return zero<cap_type>();
}

template <typename edge_t, typename cap_type = typename edge_t::capacity_type>
cap_type max_flow(graph_t<edge_t> &g, int s, int t) {
  const int V = g.size();
  cap_type flow = zero<cap_type>();
  for (;;) {
    std::vector<int> d(V, -1);
    std::queue<int> que;
    d[s] = 0;
    que.push(s);
    while (!que.empty()) {
      int v = que.front();
      que.pop();
      for (const auto &e : g[v]) {
        if (e.cap <= zero<cap_type>() || d[e.to] >= 0)
          continue;
        d[e.to] = d[v] + 1;
        que.push(e.to);
      }
    }
    if (d[t] < 0)
      return flow;
    std::vector<int> iter(V, 0);
    cap_type f;
    while ((f = augment(g, d, iter, s, t, inf<cap_type>())) > 0)
      flow += f;
  }
}
